File: lvl-1/solution.rb
class Solution

  def run(input)
    divs = find_divisors(input.length)
    divs.each do |div|
      if qualified(input, div)
        return input.length / div
      end
    end

    nil
  end

  def qualified(input, div)
    offset = 0
    while offset < div do
      return true if qualified_circle(input, div, offset)
      offset += 1
    end

    false
  end

  def qualified_circle(input, div, offset)
    base_str = get_sub_str(input, offset, div + offset)

    i = div + offset
    while i < input.length do
      upper_i = i + div
      sub_str = get_sub_str(input, i, upper_i)

      return false if sub_str != base_str
      i = upper_i
    end

    true
  end

  ###### utils

  def find_divisors(n)
    (1..n).map do |i|
      i if n % i == 0
    end.compact
  end

  def get_sub_str(input, from, to)
    to = [0, to - 1].max
    input[from..to]
  end

  def stub_circle(input, extra_len)
    input + input[0, extra_len]
  end

end

File: lvl-1/solution.py
import sys
import pdb


def answer(s):
  input = s
  divs = find_divisors(len(input))

  for div in divs:
    if qualified(input, div):
      return (len(input) / div)
  None

def qualified(input, div):
  offset = 0
  while offset < div:
    if qualified_circle(input, div, offset):
      return True
    offset += 1
  return False

def qualified_circle(input, div, offset):
  base_str = input[offset:(div + offset)]
  i = div + offset
  while i < len(input):
    next_i = i + div
    sub_str = input[i:next_i]
    if sub_str != base_str:
      return False
    i = next_i
  return True

###### utils

def find_divisors(n):
  result = []
  for i in range(1, n + 1):
    if n % i == 0:
      result.append(i)
  return result



print(answer(sys.argv[1]))
